File: cicd/api/typescript/src/pipeline.ts
/**
 * Pipeline model. Mirrors the WIA-CICD pipeline.schema.json.
 */

export const PIPELINE_NODES    = ['SOURCE', 'BUILD', 'TEST_GATE', 'SECURITY_GATE', 'CD_HANDOFF'] as const;
export const PIPELINE_TRIGGERS = ['PUSH', 'PR', 'TAG', 'MANUAL', 'SCHEDULE'] as const;
export const SECURITY_GATES    = ['SAST', 'SCA', 'DAST', 'SECRETS', 'CONTAINER', 'IAC'] as const;
export const POLICY_MODES      = ['AUDIT', 'ENFORCE'] as const;
export const DEPLOY_STRATEGIES = ['ROLLING', 'BLUE_GREEN', 'CANARY', 'PROGRESSIVE'] as const;
export const TRAFFIC_CURVE     = [5, 25, 50, 100] as const;

export type PipelineNode      = typeof PIPELINE_NODES[number];
export type PipelineTrigger   = typeof PIPELINE_TRIGGERS[number];
export type SecurityGate      = typeof SECURITY_GATES[number];
export type PolicyMode        = typeof POLICY_MODES[number];
export type DeployStrategy    = typeof DEPLOY_STRATEGIES[number];

export interface PipelineSlo {
  unit_ms?:       number;
  integration_s?: number;
  e2e_s?:         number;
}

export interface PipelineCache {
  local?:     string;
  remote?:    string;
  execution?: string;
}

export interface PipelineNodeDef {
  id:    PipelineNode;
  type:  'source' | 'build' | 'test' | 'security' | 'deliver';
  needs?: PipelineNode[];
  cache?: PipelineCache;
  slo?:   PipelineSlo;
  gates?: SecurityGate[];
  policy?: PolicyMode;
  strategy?: DeployStrategy;
  traffic_curve?: number[];
}

export interface PipelineDef {
  pipeline: {
    name:    string;
    trigger: PipelineTrigger;
    nodes:   PipelineNodeDef[];
  };
}

export interface ValidationError {
  path:    string;
  message: string;
}
export interface ValidationResult {
  valid:   boolean;
  errors:  ValidationError[];
}

/** Lightweight schema check (no external dependency). */
export function validatePipeline(def: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (!def || typeof def !== 'object' || !(def as { pipeline?: unknown }).pipeline) {
    return { valid: false, errors: [{ path: '/', message: 'missing "pipeline"' }] };
  }
  const p = (def as { pipeline: PipelineDef['pipeline'] }).pipeline;
  if (!p.name)                                      errors.push({ path: '/pipeline/name',    message: 'required' });
  if (!PIPELINE_TRIGGERS.includes(p.trigger as PipelineTrigger)) errors.push({ path: '/pipeline/trigger', message: 'invalid trigger' });
  if (!Array.isArray(p.nodes) || p.nodes.length === 0) {
    errors.push({ path: '/pipeline/nodes', message: 'at least one node required' });
  } else {
    p.nodes.forEach((n, i) => {
      if (!PIPELINE_NODES.includes(n.id as PipelineNode)) {
        errors.push({ path: `/pipeline/nodes/${i}/id`, message: `invalid id ${String(n.id)}` });
      }
      if (n.gates) {
        n.gates.forEach((g, j) => {
          if (!SECURITY_GATES.includes(g)) {
            errors.push({ path: `/pipeline/nodes/${i}/gates/${j}`, message: `invalid gate ${g}` });
          }
        });
      }
      if (n.policy   && !POLICY_MODES.includes(n.policy))           errors.push({ path: `/pipeline/nodes/${i}/policy`,   message: `invalid policy ${n.policy}` });
      if (n.strategy && !DEPLOY_STRATEGIES.includes(n.strategy))    errors.push({ path: `/pipeline/nodes/${i}/strategy`, message: `invalid strategy ${n.strategy}` });
    });
  }
  return { valid: errors.length === 0, errors };
}

/** Topological sort over node `needs:` edges. Throws on cycle. */
export function topoSort(def: PipelineDef): PipelineNode[] {
  const nodes  = def.pipeline.nodes;
  const inDeg  = new Map<PipelineNode, number>();
  const edges  = new Map<PipelineNode, PipelineNode[]>();
  nodes.forEach(n => { inDeg.set(n.id, 0); edges.set(n.id, []); });
  nodes.forEach(n => (n.needs ?? []).forEach(prev => {
    inDeg.set(n.id, (inDeg.get(n.id) ?? 0) + 1);
    edges.get(prev)?.push(n.id);
  }));
  const ready: PipelineNode[] = [];
  inDeg.forEach((d, k) => { if (d === 0) ready.push(k); });
  ready.sort();
  const order: PipelineNode[] = [];
  while (ready.length > 0) {
    const cur = ready.shift() as PipelineNode;
    order.push(cur);
    edges.get(cur)?.forEach(next => {
      const left = (inDeg.get(next) ?? 0) - 1;
      inDeg.set(next, left);
      if (left === 0) { ready.push(next); ready.sort(); }
    });
  }
  if (order.length !== nodes.length) throw new Error('cycle detected in pipeline DAG');
  return order;
}

File: cicd/api/rust/src/pipeline.rs
//! Pipeline model. Mirrors `schemas/pipeline.schema.json`.

use std::collections::{BTreeMap, BTreeSet};

/// Five canonical DAG node identifiers.
pub const PIPELINE_NODES: [&str; 5] = ["SOURCE", "BUILD", "TEST_GATE", "SECURITY_GATE", "CD_HANDOFF"];

/// Recognised pipeline triggers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Trigger { /// push to a tracked branch
    Push,
    /// pull request open / sync
    Pr,
    /// tag publish
    Tag,
    /// manual dispatch
    Manual,
    /// schedule / cron
    Schedule
}

impl Trigger {
    /// Stringify into the spec ENUM value.
    pub fn as_str(&self) -> &'static str {
        match self {
            Trigger::Push => "PUSH",
            Trigger::Pr => "PR",
            Trigger::Tag => "TAG",
            Trigger::Manual => "MANUAL",
            Trigger::Schedule => "SCHEDULE",
        }
    }
}

/// Recognised security gates.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum SecurityGate { /// static application security testing
    Sast,
    /// software composition analysis
    Sca,
    /// dynamic application security testing
    Dast,
    /// secrets scan
    Secrets,
    /// container image scan
    Container,
    /// infrastructure as code scan
    Iac
}

impl SecurityGate {
    /// Stringify.
    pub fn as_str(&self) -> &'static str {
        match self {
            SecurityGate::Sast => "SAST",
            SecurityGate::Sca => "SCA",
            SecurityGate::Dast => "DAST",
            SecurityGate::Secrets => "SECRETS",
            SecurityGate::Container => "CONTAINER",
            SecurityGate::Iac => "IAC",
        }
    }
}

/// Policy mode.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum PolicyMode { /// observe only
    Audit,
    /// block on policy fail
    Enforce
}

/// Deployment strategy.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DeployStrategy { /// rolling update
    Rolling,
    /// blue-green
    BlueGreen,
    /// canary
    Canary,
    /// progressive (canary + analysis)
    Progressive
}

/// Functional class of a pipeline node.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NodeType { /// SCM source
    Source,
    /// build / package
    Build,
    /// test gate
    Test,
    /// security gate
    Security,
    /// continuous delivery handoff
    Deliver
}

/// A single pipeline node.
#[derive(Debug, Clone)]
pub struct Node {
    /// Canonical identifier (member of `PIPELINE_NODES`).
    pub id:        String,
    /// Functional class.
    pub kind:      NodeType,
    /// IDs of upstream nodes that must complete before this one.
    pub needs:     Vec<String>,
    /// Security gates evaluated by this node (Security only).
    pub gates:     Vec<SecurityGate>,
    /// Policy mode (Security only).
    pub policy:    Option<PolicyMode>,
    /// Deploy strategy (Deliver only).
    pub strategy:  Option<DeployStrategy>,
    /// Traffic curve percentages (Deliver only).
    pub traffic_curve: Vec<u8>,
}

impl Node {
    /// Construct a node with no dependencies.
    pub fn new(id: &str, kind: NodeType) -> Self {
        Self { id: id.to_string(), kind, needs: vec![], gates: vec![], policy: None, strategy: None, traffic_curve: vec![] }
    }
    /// Builder: set dependencies.
    pub fn needs(mut self, deps: &[&str]) -> Self {
        self.needs = deps.iter().map(|d| d.to_string()).collect(); self
    }
}

/// A pipeline.
#[derive(Debug, Clone)]
pub struct Pipeline {
    /// Pipeline name.
    pub name:    String,
    /// Trigger.
    pub trigger: Trigger,
    /// Nodes.
    pub nodes:   Vec<Node>,
}

/// Validation error.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValidationError {
    /// JSON pointer-like path.
    pub path:    String,
    /// Human-readable message.
    pub message: String,
}

/// Validate a pipeline against the WIA-CICD spec ENUMs.
pub fn validate(p: &Pipeline) -> Result<(), Vec<ValidationError>> {
    let mut errs = Vec::new();
    if p.name.is_empty() { errs.push(ValidationError { path: "/name".into(), message: "required".into() }); }
    if p.nodes.is_empty() { errs.push(ValidationError { path: "/nodes".into(), message: "at least one node required".into() }); }
    let valid_ids: BTreeSet<&str> = PIPELINE_NODES.iter().copied().collect();
    for (i, n) in p.nodes.iter().enumerate() {
        if !valid_ids.contains(n.id.as_str()) {
            errs.push(ValidationError { path: format!("/nodes/{}/id", i), message: format!("invalid id {}", n.id) });
        }
    }
    if errs.is_empty() { Ok(()) } else { Err(errs) }
}

/// Topological sort. Returns the linearised execution order or an error on cycle.
pub fn topo_sort(p: &Pipeline) -> Result<Vec<String>, &'static str> {
    let mut in_deg: BTreeMap<&str, usize>     = BTreeMap::new();
    let mut edges:  BTreeMap<&str, Vec<&str>> = BTreeMap::new();
    for n in &p.nodes { in_deg.insert(&n.id, 0); edges.entry(&n.id).or_default(); }
    for n in &p.nodes {
        for prev in &n.needs {
            *in_deg.entry(&n.id).or_insert(0) += 1;
            edges.entry(prev.as_str()).or_default().push(&n.id);
        }
    }
    let mut ready: Vec<&str> = in_deg.iter().filter(|(_, &d)| d == 0).map(|(k, _)| *k).collect();
    ready.sort();
    let mut order: Vec<String> = Vec::new();
    while let Some(cur) = ready.first().copied() {
        ready.remove(0);
        order.push(cur.to_string());
        if let Some(nexts) = edges.get(cur) {
            for next in nexts.clone() {
                let entry = in_deg.entry(next).or_insert(1);
                *entry = entry.saturating_sub(1);
                if *entry == 0 { ready.push(next); ready.sort(); }
            }
        }
    }
    if order.len() != p.nodes.len() { return Err("cycle detected"); }
    Ok(order)
}
